# Setup


In [1]:
import pandas as pd
import json

# METADATA

1. `movie.metadata.tsv.gz` [3.4 M]


Metadata for 81,741 movies, extracted from the Noverber 4, 2012 dump of Freebase.  Tab-separated; columns:

1. Wikipedia movie ID
2. Freebase movie ID
3. Movie name
4. Movie release date
5. Movie box office revenue
6. Movie runtime
7. Movie languages (Freebase ID:name tuples)
8. Movie countries (Freebase ID:name tuples)
9. Movie genres (Freebase ID:name tuples)



2. `character.metadata.tsv.gz` [14 M]

Metadata for 450,669 characters aligned to the movies above, extracted from the Noverber 4, 2012 dump of Freebase.  Tab-separated; columns:

1. Wikipedia movie ID
2. Freebase movie ID
3. Movie release date
4. Character name
5. Actor date of birth
6. Actor gender
7. Actor height (in meters)
8. Actor ethnicity (Freebase ID)
9. Actor name
10. Actor age at movie release
11. Freebase character/actor map ID
12. Freebase character ID
13. Freebase actor ID

# Importing the datasets


In [2]:
movies_path = "data/movie.metadata.tsv"
character_path = "data/character.metadata.tsv"

movie_column_names = ["WikiID", "FreeID", "Title", "RelDate", "Revenue", "Runtime", "Languages", "Countries", "Genres"]
character_column_names = ["WikiID", "FreeID", "MovieRelDate", "CharName", "DOB", "Gender", "Height", "Ethnicity", "Actor", "Age", "FreeMapID", "FreeCharID", "FreeActorID"]

movies = pd.read_csv(movies_path, sep='\t', header=None, names=movie_column_names)
characters = pd.read_csv(character_path, sep='\t', header=None, names=character_column_names)

# Looking at the data

In [3]:
display(movies.head())
display(characters.head())

WikiID      FreeID                                              Title  \
0    975900   /m/03vyhn                                     Ghosts of Mars   
1   3196793   /m/08yl5d  Getting Away with Murder: The JonBenét Ramsey ...   
2  28463795  /m/0crgdbh                                        Brun bitter   
3   9363483  /m/0285_cd                                   White Of The Eye   
4    261236   /m/01mrr1                                  A Woman in Flames   

      RelDate     Revenue  Runtime                           Languages  \
0  2001-08-24  14010832.0     98.0  {"/m/02h40lc": "English Language"}   
1  2000-02-16         NaN     95.0  {"/m/02h40lc": "English Language"}   
2        1988         NaN     83.0  {"/m/05f_3": "Norwegian Language"}   
3        1987         NaN    110.0  {"/m/02h40lc": "English Language"}   
4        1983         NaN    106.0   {"/m/04306rv": "German Language"}   

                                   Countries  \
0  {"/m/09c7w0": "United States of America"}   
1  {"/m/09c7w0": "United States of America"}   
2                     {"/m/05b4w": "Norway"}   
3             {"/m/07ssc": "United Kingdom"}   
4                    {"/m/0345h": "Germany"}   

                                              Genres  
0  {"/m/01jfsb": "Thriller", "/m/06n90": "Science...  
1  {"/m/02n4kr": "Mystery", "/m/03bxz7": "Biograp...  
2  {"/m/0lsxr": "Crime Fiction", "/m/07s9rl0": "D...  
3  {"/m/01jfsb": "Thriller", "/m/0glj9q": "Erotic...  
4                            {"/m/07s9rl0": "Drama"}

WikiID     FreeID MovieRelDate                    CharName         DOB  \
0  975900  /m/03vyhn   2001-08-24                    Akooshay  1958-08-26   
1  975900  /m/03vyhn   2001-08-24  Lieutenant Melanie Ballard  1974-08-15   
2  975900  /m/03vyhn   2001-08-24         Desolation Williams  1969-06-15   
3  975900  /m/03vyhn   2001-08-24          Sgt Jericho Butler  1967-09-12   
4  975900  /m/03vyhn   2001-08-24             Bashira Kincaid  1977-09-25   

  Gender  Height   Ethnicity               Actor   Age   FreeMapID  \
0      F   1.620         NaN      Wanda De Jesus  42.0  /m/0bgchxw   
1      F   1.780  /m/044038p  Natasha Henstridge  27.0   /m/0jys3m   
2      M   1.727     /m/0x67            Ice Cube  32.0   /m/0jys3g   
3      M   1.750         NaN       Jason Statham  33.0  /m/02vchl6   
4      F   1.650         NaN         Clea DuVall  23.0  /m/02vbb3r   

   FreeCharID FreeActorID  
0  /m/0bgcj3x  /m/03wcfv7  
1  /m/0bgchn4   /m/0346l4  
2  /m/0bgchn_  /m/01vw26l  
3  /m/0bgchnq   /m/034hyc  
4  /m/0bgchp9   /m/01y9xg



*   Note : `RelDate`doesn't have a standard format



### Exploring length of the two datasets

In [4]:
n_mov = len(movies)
n_char = len(characters)
print('Number of rows in the movie.metadata dataset :', n_mov)
print('Number of rows in the movie.metadata dataset :', n_char)

Number of rows in the movie.metadata dataset : 81741
Number of rows in the movie.metadata dataset : 450669


# Cleaning Dictionaries
Columns `Languages`, `Countries`, `Genres` contain wikipedia ID + actual name for each entry. For the moment, keeping just the name for the sake of clarity

In [5]:
def extract_values(column):
  values = []
  column=json.loads(column)
  for key in column:
    values.append(column[key])
  return values

movies.Languages = movies.Languages.apply(extract_values)
movies.Countries = movies.Countries.apply(extract_values)
movies.Genres = movies.Genres.apply(extract_values)

In [6]:
display(movies[["Languages", "Countries", "Genres"]].head(2))

Languages                   Countries  \
0  [English Language]  [United States of America]   
1  [English Language]  [United States of America]   

                                              Genres  
0  [Thriller, Science Fiction, Horror, Adventure,...  
1   [Mystery, Biographical film, Drama, Crime Drama]

# Duplicates

In [7]:
for column in movies.columns:
  duplicated_rows = movies[column].duplicated().sum()
  print("{} has {} duplicated rows".format(column, duplicated_rows))

WikiID has 0 duplicated rows
FreeID has 0 duplicated rows
Title has 6263 duplicated rows
RelDate has 61351 duplicated rows
Revenue has 74378 duplicated rows
Runtime has 81143 duplicated rows
Languages has 79924 duplicated rows
Countries has 79617 duplicated rows
Genres has 57924 duplicated rows


`WikiID`& `FreeID`-> Good

`Title`-> Bit concerning -> Check with Runtime


> ~~Runtime not enough. There are :~~

1.   ~~Movies with same runtime + Title but actually different~~
2.   ~~Movies with same runtime + Title but actually the same~~





In [8]:
movies[movies.Title=="Hunting Season"]

WikiID      FreeID           Title     RelDate     Revenue  Runtime  \
62836  29666067  /m/0fphzrf  Hunting Season  1010-12-02  12160978.0    140.0   

                                  Languages Countries  \
62836  [Turkish Language, English Language]  [Turkey]   

                                          Genres  
62836  [Crime Fiction, Mystery, Drama, Thriller]

Wrong input for release date (1010->2010)

In [9]:
movies[movies.Title=="Harlow"]

WikiID      FreeID   Title     RelDate    Revenue  Runtime  \
623    3670013   /m/09thsq  Harlow  1965-06-23  1000000.0    109.0   
1223  27171821  /m/0bwklv0  Harlow        1965        NaN    109.0   

               Languages                   Countries  \
623   [English Language]  [United States of America]   
1223                  []  [United States of America]   

                                                 Genres  
623   [Biographical film, Biography, Drama, Black-an...  
1223                                [Biographical film]

Actually these are 2 different movies BUT first movie `WikiID`=3670013 has wrong rutime (correct =125).

So far couldn't find an example of a duplicated movie + No duplicated Wiki/Freebase IDs -> Lets trust the dataset 🙃

# Missing values

### Check for missing entries :
In which columns are they present, and for those columns : what is the percentage how missing entries.

In [10]:
movies.isna().any()

WikiID       False
FreeID       False
Title        False
RelDate       True
Revenue       True
Runtime       True
Languages    False
Countries    False
Genres       False
dtype: bool

In [11]:
print('Percentage of missing entries in the movie dataset:\n', 100*movies[['RelDate', 'Revenue', 'Runtime']].isna().sum() / n_mov)

Percentage of missing entries in the movie dataset:
 RelDate     8.443743
Revenue    89.722416
Runtime    25.018045
dtype: float64


There are 90% of revenues that are non specified. We probably won't be able to use this feature.  
8% of the release dates and 25% of the runtimes are missing, we can fill them if we find the correct ones.

In [12]:
characters.isna().any()

WikiID          False
FreeID          False
MovieRelDate     True
CharName         True
DOB              True
Gender           True
Height           True
Ethnicity        True
Actor            True
Age              True
FreeMapID       False
FreeCharID       True
FreeActorID      True
dtype: bool

In [14]:
print('Percentage of missing entries in the character dataset:\n', 100*characters[['FreeID', 'MovieRelDate', 'CharName',
                                                                                   'DOB', 'Gender', 'Height', 'Ethnicity',
                                                                                   'Actor', 'Age', 'FreeMapID', 'FreeCharID',
                                                                                   'FreeActorID']].isna().sum() / n_char)

Percentage of missing entries in the character dataset:
 FreeID           0.000000
MovieRelDate     2.217814
CharName        57.220488
DOB             23.552763
Gender          10.120288
Height          65.645740
Ethnicity       76.466542
Actor            0.272484
Age             35.084064
FreeMapID        0.000000
FreeCharID      57.218269
FreeActorID      0.180842
dtype: float64


# Wikidata API

To begin with there are some features like `Èthnicity` and `FreeCharId`that only have the Id of the Freebase database whose API is currently depreciated. To solve this problem we use the [Freebase/Wikidata Mappings](https://developers.google.com/freebase#freebase-wikidata-mappings) which as the name suggests maps the given Freebase Ids to WikiData.

- Note : The data has been created based on the Wikidata-Dump of October 28, 2013, and contains only those links that have at least two common Wikipedia-Links and not a single disagreeing Wikipedia-Link. Since the movies/characters dataset were gathered during that time too there is no time conflict.

In [15]:
mappings_path = "data/fb2w.nt.gz"
maps = pd.read_csv(mappings_path, sep='\t', header=None, skiprows=4, names=["Freebase", "W3", "Wiki"])
display(maps.head())

Freebase  \
0  <http://rdf.freebase.com/ns/m.0695j>   
1  <http://rdf.freebase.com/ns/m.05nrg>   
2   <http://rdf.freebase.com/ns/m.0jgd>   
3  <http://rdf.freebase.com/ns/m.0d_23>   
4  <http://rdf.freebase.com/ns/m.04g7d>   

                                       W3  \
0  <http://www.w3.org/2002/07/owl#sameAs>   
1  <http://www.w3.org/2002/07/owl#sameAs>   
2  <http://www.w3.org/2002/07/owl#sameAs>   
3  <http://www.w3.org/2002/07/owl#sameAs>   
4  <http://www.w3.org/2002/07/owl#sameAs>   

                                       Wiki  
0  <http://www.wikidata.org/entity/Q6718> .  
1   <http://www.wikidata.org/entity/Q538> .  
2   <http://www.wikidata.org/entity/Q414> .  
3  <http://www.wikidata.org/entity/Q2537> .  
4   <http://www.wikidata.org/entity/Q315> .

We only need the Id and not the whole website link so we transform columns in the following way :
- Freebase : <http://rdf.freebase.com/ns/m.CODE_ID> -> /m/CODE_ID
- Wiki : <http://www.wikidata.org/entity/CODE_ID> -> CODE_ID

In [15]:
"""
Transforms the Freebase website to Freebase ID
"""
def elim_freebase(web):
    return "/m/"+web[30:-1]

"""
Transforms the Freebase website to Freebase ID
"""
def elim_wiki(web):
    return web[32:-3]

In [16]:
maps["free"] = maps.Freebase.apply(elim_freebase)
maps["wiki"] = maps.Wiki.apply(elim_wiki)

# We don't need the rest of the columns so we drop them 
maps.drop(columns = ["Freebase", "W3", "Wiki"], inplace=True)
maps.head()

free   wiki
0  /m/0695j  Q6718
1  /m/05nrg   Q538
2   /m/0jgd   Q414
3  /m/0d_23  Q2537
4  /m/04g7d   Q315

## Ethnicity labels

Firstly, we get the Wikidata IDs based on the mapping we created :

In [17]:
characters = characters.merge(maps, how="left", left_on="Ethnicity", right_on="free")
characters.drop(columns=["free"], inplace=True)
characters.rename(columns={"wiki":"Ethnicity_W"}, inplace=True)
display(characters)

WikiID      FreeID MovieRelDate                    CharName  \
0         975900   /m/03vyhn   2001-08-24                    Akooshay   
1         975900   /m/03vyhn   2001-08-24  Lieutenant Melanie Ballard   
2         975900   /m/03vyhn   2001-08-24         Desolation Williams   
3         975900   /m/03vyhn   2001-08-24          Sgt Jericho Butler   
4         975900   /m/03vyhn   2001-08-24             Bashira Kincaid   
...          ...         ...          ...                         ...   
450727    913762   /m/03pcrp   1992-05-21                      Elensh   
450728    913762   /m/03pcrp   1992-05-21                      Hibiki   
450729  28308153  /m/0cp05t9         1957                         NaN   
450730  28308153  /m/0cp05t9         1957                         NaN   
450731  28308153  /m/0cp05t9         1957                         NaN   

               DOB Gender  Height   Ethnicity               Actor   Age  \
0       1958-08-26      F   1.620         NaN      Wanda De Jesus  42.0   
1       1974-08-15      F   1.780  /m/044038p  Natasha Henstridge  27.0   
2       1969-06-15      M   1.727     /m/0x67            Ice Cube  32.0   
3       1967-09-12      M   1.750         NaN       Jason Statham  33.0   
4       1977-09-25      F   1.650         NaN         Clea DuVall  23.0   
...            ...    ...     ...         ...                 ...   ...   
450727     1970-05      F     NaN         NaN  Dorothy Elias-Fahn   NaN   
450728  1965-04-12      M     NaN         NaN       Jonathan Fahn  27.0   
450729  1941-11-18      M   1.730   /m/02w7gg      David Hemmings  15.0   
450730         NaN    NaN     NaN         NaN    Roberta Paterson   NaN   
450731         NaN    NaN     NaN         NaN         John Rogers   NaN   

         FreeMapID  FreeCharID FreeActorID Ethnicity_W  
0       /m/0bgchxw  /m/0bgcj3x  /m/03wcfv7         NaN  
1        /m/0jys3m  /m/0bgchn4   /m/0346l4         NaN  
2        /m/0jys3g  /m/0bgchn_  /m/01vw26l      Q49085  
3       /m/02vchl6  /m/0bgchnq   /m/034hyc         NaN  
4       /m/02vbb3r  /m/0bgchp9   /m/01y9xg         NaN  
...            ...         ...         ...         ...  
450727  /m/0kr406c  /m/0kr406h   /m/0b_vcv         NaN  
450728  /m/0kr405_  /m/0kr4090   /m/0bx7_j         NaN  
450729  /m/0g8ngmc         NaN   /m/022g44      Q42406  
450730  /m/0g8ngmj         NaN  /m/0g8ngmm         NaN  
450731  /m/0g8ngmw         NaN  /m/0btz19d         NaN  

[450732 rows x 14 columns]

As we can see the mapping is not complete : Some Freebase IDs do not correspond to any Wikidata ID. 

In [18]:
unmapped_ethnicities = characters.Ethnicity_W.isna().sum() - characters.Ethnicity.isna().sum()
print("There are {} instances of ethnicities that we could not map from Freebase to Wikidata".format(unmapped_ethnicities))
# Instances of ethnicities -> Can we replace by number of actors ethnicities? Actors with different initial ethnicities ?

There are 12992 instances of ethnicities that we could not map from Freebase to Wikidata


In [19]:
from wikidata.client import Client

"""
Given a Wikidata ID (eg. "Q49085") returns the corresponding label 

Using this function row by row (pd.apply method) takes enormous time -> improvement? For ethnicities its ok as they're only 350 dinstinct ones
"""
def get_wikidata_label(wikidata_id):
    client = Client()
    entity = client.get(wikidata_id, load=True)
    return str(entity.label)

In [20]:
# No official comment haha this will probably change (Takes 2-3 minutes to complete). 
# Result is map_Ethnicity_W_to_Ethnicity_name a map between Ethnicity wikidata ID and Ethnicity label
map_Ethnicity_W_to_Ethnicity_name = pd.DataFrame(characters.Ethnicity_W.unique(), columns=["Ethnicity_W"]).dropna()
map_Ethnicity_W_to_Ethnicity_name["Ethnicity_name"] = map_Ethnicity_W_to_Ethnicity_name.Ethnicity_W.apply(get_wikidata_label)

In [21]:
characters = characters.merge(map_Ethnicity_W_to_Ethnicity_name, how="left", left_on="Ethnicity_W", right_on="Ethnicity_W").drop(columns= ["Ethnicity", "Ethnicity_W"])
display(characters)

WikiID      FreeID MovieRelDate                    CharName  \
0         975900   /m/03vyhn   2001-08-24                    Akooshay   
1         975900   /m/03vyhn   2001-08-24  Lieutenant Melanie Ballard   
2         975900   /m/03vyhn   2001-08-24         Desolation Williams   
3         975900   /m/03vyhn   2001-08-24          Sgt Jericho Butler   
4         975900   /m/03vyhn   2001-08-24             Bashira Kincaid   
...          ...         ...          ...                         ...   
450727    913762   /m/03pcrp   1992-05-21                      Elensh   
450728    913762   /m/03pcrp   1992-05-21                      Hibiki   
450729  28308153  /m/0cp05t9         1957                         NaN   
450730  28308153  /m/0cp05t9         1957                         NaN   
450731  28308153  /m/0cp05t9         1957                         NaN   

               DOB Gender  Height               Actor   Age   FreeMapID  \
0       1958-08-26      F   1.620      Wanda De Jesus  42.0  /m/0bgchxw   
1       1974-08-15      F   1.780  Natasha Henstridge  27.0   /m/0jys3m   
2       1969-06-15      M   1.727            Ice Cube  32.0   /m/0jys3g   
3       1967-09-12      M   1.750       Jason Statham  33.0  /m/02vchl6   
4       1977-09-25      F   1.650         Clea DuVall  23.0  /m/02vbb3r   
...            ...    ...     ...                 ...   ...         ...   
450727     1970-05      F     NaN  Dorothy Elias-Fahn   NaN  /m/0kr406c   
450728  1965-04-12      M     NaN       Jonathan Fahn  27.0  /m/0kr405_   
450729  1941-11-18      M   1.730      David Hemmings  15.0  /m/0g8ngmc   
450730         NaN    NaN     NaN    Roberta Paterson   NaN  /m/0g8ngmj   
450731         NaN    NaN     NaN         John Rogers   NaN  /m/0g8ngmw   

        FreeCharID FreeActorID     Ethnicity_name  
0       /m/0bgcj3x  /m/03wcfv7                NaN  
1       /m/0bgchn4   /m/0346l4                NaN  
2       /m/0bgchn_  /m/01vw26l  African Americans  
3       /m/0bgchnq   /m/034hyc                NaN  
4       /m/0bgchp9   /m/01y9xg                NaN  
...            ...         ...                ...  
450727  /m/0kr406h   /m/0b_vcv                NaN  
450728  /m/0kr4090   /m/0bx7_j                NaN  
450729         NaN   /m/022g44     English people  
450730         NaN  /m/0g8ngmm                NaN  
450731         NaN  /m/0btz19d                NaN  

[450732 rows x 13 columns]

In [22]:
characters.Ethnicity_name.value_counts()[:10] 
# Should we even continue with these ethnicities ? These results aren't affected by the lack of mappings between Freebase and Wiki

Indians              16351
African Americans    10159
Jewish people        10121
English people        6165
Irish Americans       3857
Italian Americans     2956
white Americans       2250
Scottish American     2223
Irish people          1637
French                1350
Name: Ethnicity_name, dtype: int64

# DATA

This part deals with plot summaries data.

1. `plot_summaries.txt.gz` [29 M]

Plot summaries of 42,306 movies extracted from the November 2, 2012 dump of English-language Wikipedia.  Each line contains the Wikipedia movie ID (which indexes into `movie.metadata.tsv`) followed by the summary.


2. `corenlp_plot_summaries.tar.gz` [628 M, separate download]

The plot summaries from above, run through the Stanford CoreNLP pipeline (tagging, parsing, NER and coref). Each filename begins with the Wikipedia movie ID (which indexes into `movie.metadata.tsv`).


# TEST DATA

1. `tvtropes.clusters.txt`

72 character types drawn from tvtropes.com, along with 501 instances of those types.  The ID field indexes into the Freebase character/actor map ID in `character.metadata.tsv`.

2. `name.clusters.txt`


970 unique character names used in at least two different movies, along with 2,666 instances of those types.  The ID field indexes into the Freebase character/actor map ID in `character.metadata.tsv`.


In [23]:
tvtropes_path = "data/tvtropes.clusters.txt"
name_path = "data/name.clusters.txt"

##### Mapping: put clusters info into character dataset:

---